src/dataflow/src/render/reduce.rs

```diff
@@ -606,3 +618,86 @@ fn is_min_or_max(func: &AggregateFunc) -> bool {
         _ => false,
     }
 }
+
+fn is_max(func: &AggregateFunc) -> bool {
+    match func {
+        AggregateFunc::MaxInt32
+        | AggregateFunc::MaxInt64
+        | AggregateFunc::MaxFloat32
+        | AggregateFunc::MaxFloat64
+        | AggregateFunc::MaxDecimal
+        | AggregateFunc::MaxBool
+        | AggregateFunc::MaxString
+        | AggregateFunc::MaxDate
+        | AggregateFunc::MaxTimestamp
+        | AggregateFunc::MaxTimestampTz => true,
+        _ => false,
+    }
+}
+
+/// Monoids for in-place compaction of monotonic streams.
+pub mod monoids {
+    use repr::{Datum, Row};
+    use serde::{Deserialize, Serialize};
+
+    #[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Clone, Serialize, Deserialize, Hash)]
+    pub struct MinMonoid {
+        pub value: Row,
+    }
+
+    use differential_dataflow::difference::Semigroup;
+    use std::ops::AddAssign;
+
+    impl<'a> AddAssign<&'a Self> for MinMonoid {
+        fn add_assign(&mut self, rhs: &'a Self) {
+            let swap = {
+                let lhs_val = self.value.unpack_first();
+                let rhs_val = rhs.value.unpack_first();
+                // Datum::Null is the identity, not a small element.
+                match (lhs_val, rhs_val) {
```

Comment:
if the row has more than one datum - would it be right to then compare the second ones if the first ones are equal?

it seems like the row must have exactly one Datum for the things we want to use this for, but just sanity checking my understanding here

Reply:
Yup, the intent here is that these contain single values. We could pack multiple `max` statements together into one collection, and then multiple `min` statements together too.

Reply:
Adding some comments to better explain "assumptions"